### PR TITLE
Combine Multi_Pin_Conduction and Multi_Pin_Subchannel

### PR DESCRIPTION
**Pull request by aprilnovak**
While Nek solves for both the fluid and solid phases together, the `smrt` surrogate heat solvers solve the fluid and solid phases separately in the `Multi_Pin_Subchannel` and `Multi_Pin_Conduction` classes. To have unified base class drivers, these two classes are combined here into a new class, `SurrogateHeatFluidDriver` (if anyone has a better naming idea I'm all ears).

This better reflects that the T/H solve is performed over _both_ phases; the solution in the fluid and solid is nonlinearly coupled due to the convective BC, though in this application this is represented with explicit coupling without iteration within the T/H solver.

This is a first step towards adapting `SurrogateHeatFluidDriver` to inherit from `HeatFluidsDriver`.

### CODE REVIEW

**Comment by aprilnovak** [comment]
In this surrogate T/H driver, we will need to somehow combine the fluid temperature and solid temperature vectors into a single vector (since this is how Nek represents the temperatures). Does anyone have a preference on how this should be done? Simply append the solid temperatures after the fluid temperatures, with a `fluid_mask` of a bunch of 1's followed by a bunch of 0's?

File: src/smrt/surrogate_heat_fluid_driver.cpp [hunk 1]
```diff
@@ -69,10 +75,16 @@ void Multi_Pin_Subchannel::generate_arrays()
   d_pin_powers.resize(pins_x * pins_y * d_Nz);
 }
 
+void SurrogateHeatFluidDriver::solve(const std::vector<double>& powers)
```
Comment (paulromano) [inline src/smrt/surrogate_heat_fluid_driver.cpp]: I don't see a declaration for this in the header file.

File: src/smrt/surrogate_heat_fluid_driver.cpp [hunk 2]
```diff
@@ -156,9 +168,57 @@ void Multi_Pin_Subchannel::solve(const std::vector<double>& powers)
   }
 }
 
+void SurrogateHeatFluidDriver::solve_solid(const std::vector<double>& power,
```
Comment (paulromano) [inline src/smrt/surrogate_heat_fluid_driver.cpp]: `solve_heat` would match the class name better

File: include/smrt/surrogate_heat_fluid_driver.h [hunk 3]
```diff
@@ -51,7 +53,11 @@ class Multi_Pin_Subchannel {
   // Mass flow rate (kg/s) in each channel
   std::vector<double> d_mdots;
 
-  SP_Pin_Subchannel d_pin_subchannel;
+  //! subchannel solver for a single channel
+  std::shared_ptr<Single_Pin_Subchannel> d_pin_subchannel;
+
+  //! heat conduction solver for a single rod
+  std::shared_ptr<Single_Pin_Conduction> d_pin_conduction;
```
[the remark this answers is not in the document]
Comment (aprilnovak) [reply include/smrt/surrogate_heat_fluid_driver.h]: `unique_ptr` is probably fine, I just formed these files mostly by copying and pasting the two separate files together, and this is how things were before. I'll update

**Comment by paulromano** [comment]
All looks good to me, but I'll wait to merge until we have confirmation that it actually compiles.

File: include/smrt/surrogate_heat_fluid_driver.h [hunk 4]
```diff
@@ -83,14 +94,21 @@ class Multi_Pin_Subchannel {
     Expects(iy < d_Ny);
     return ix + d_Nx * iy;
   }
+
+  // Solve the fluid flow equations
+  void solve_fluid(const std::vector<double>& power);
+
+  // Solve the solid equations
+  void solve_heat(const std::vector<double>& power,
+    const std::vector<double>& channel_temp, std::vector<double>& fuel_temp)
```
Comment (sphamil) [inline include/smrt/surrogate_heat_fluid_driver.h]: Missing `;`

File: include/smrt/Multiphysics_Driver.h [hunk 5]
```diff
@@ -8,10 +8,10 @@
 #include "Teuchos_RCP.hpp"
 
 #include "Assembly_Model.h"
-#include "Multi_Pin_Conduction.h"
-#include "Multi_Pin_Subchannel.h"
 #include "Neutronics_Solver.h"
 
+#include "surrogate_heat_fluids_driver.h"
```
Comment (sphamil) [inline include/smrt/Multiphysics_Driver.h]: Should be `surrogate_heat_fluid_driver` (no `s`).

**Comment by sphamil** [comment]
Need to remove `src/smrt/Multi_Pin_{Conduction,Subchannel}.cpp` from `CMakeLists.txt`

**Comment by aprilnovak** [comment]
Thanks for the checks @sphamil!

**Comment by sphamil** [comment]
I think this is good to be merged on my account.

**Pull request merged**